src/main/java/com/hw/bookstore/specification/BookSpecification.java

```diff
@@ -0,0 +1,48 @@
+package com.hw.bookstore.specification;
+
+import com.hw.bookstore.domain.entity.Book;
+import com.hw.bookstore.domain.entity.Book_;
+import com.hw.bookstore.dto.BookSearchParamsRequestDto;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import java.util.ArrayList;
+import lombok.AllArgsConstructor;
+import org.springframework.data.jpa.domain.Specification;
+
+@AllArgsConstructor(staticName = "of")
+public class BookSpecification implements Specification<Book> {
+
+    private final BookSearchParamsRequestDto requestDto;
+
+    @Override
+    public Predicate toPredicate(Root<Book> root, CriteriaQuery<?> cq, CriteriaBuilder cb) {
+        var predicates = new ArrayList<Predicate>();
+
+        if (requestDto.authors() != null && !requestDto.authors().isEmpty()) {
+            predicates.add(root.get(Book_.AUTHOR).in(requestDto.authors()));
+        }
+
+        if (requestDto.titles() != null && !requestDto.titles().isEmpty()) {
+            predicates.add(root.get(Book_.TITLE).in(requestDto.titles()));
+        }
+
+        if (requestDto.isbn() != null) {
+            predicates.add(cb.equal(root.get(Book_.ISBN), requestDto.isbn()));
+        }
+
+        if (requestDto.lowerPrice() != null && requestDto.upperPrice() != null) {
+            predicates.add(cb.between(
+                    root.get(Book_.PRICE), requestDto.lowerPrice(), requestDto.upperPrice()));

```

Comment:
I guess this conditional is redundant. It is the same that both Predicate - `greaterThanOrEqualTo` and `lessThanOrEqualTo`

Reply:
Some of this field could be null, so this condition will be valid.